### PR TITLE
feat(FN-2420): ukef utilisation report email notifications to multiple addresses

### PR DESCRIPTION
**Pull request by RLavender98**
## Introduction :pencil2:
Add support for sending utilisation report upload emails to multiple email addresses.

## Resolution :heavy_check_mark:
- Use a list of emails in the env var for UKEF report email notification recipients
- Send utilisation report uploaded notifications to all emails in the list

## Miscellaneous :heavy_plus_sign:
- Refactor the upload controller to extract some of the logic into separate services
- Add some unit tests for the controller and created services
- Don't fail upload request if sending notification email to UKEF fails

### CODE REVIEW

File: portal-api/src/v1/services/utilisation-report/email-service.ts
Comment (RLavender98) [inline portal-api/src/v1/services/utilisation-report/email-service.ts:13]: This requires emails to be provided as: 
`UKEF_GEF_REPORTING_EMAIL_RECIPIENT=["email1", "email2"]`
The env var is just read as a string so we can change this to a different form if we could like, just anything we can then parse and turn into an array

File: portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js
Comment (RLavender98) [inline portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js:31]: This catching here is a change to functionality, it just felt weird to be showing the user an error page when this emails fails to send given they are not told it was sending it in the first place and otherwise the upload went fine. But if we don't want this I can change this back. I thought about changing the other email to also fail quietly from the user's perspective but left it since we display the email addresses on the page so that would have required more thinking.

File: portal-api/src/v1/services/utilisation-report/azure-file-service.js
Comment (RLavender98) [inline portal-api/src/v1/services/utilisation-report/azure-file-service.js:1]: I wasn't too sure on the best place to put this and the email-service and indeed their names. Settled for a services folder, with a `utilisation-reports` folder then `-service` file names. Keen to know if there is a better pattern I could use.

File: portal-api/src/v1/services/utilisation-report/email-service.ts
Comment (JoshBinns2000) [inline portal-api/src/v1/services/utilisation-report/email-service.ts:39]: Missing `@param`

File: portal-api/src/v1/services/utilisation-report/email-service.ts
Comment (JoshBinns2000) [inline portal-api/src/v1/services/utilisation-report/email-service.ts:53]: Missing `@param`s

File: portal-api/src/v1/validation/utilisation-report/utilisation-report-upload-validator.ts
Comment (JoshBinns2000) [inline portal-api/src/v1/validation/utilisation-report/utilisation-report-upload-validator.ts:5]: This feels to me like it is doing too much for one function 🤔

File: portal-api/src/v1/services/utilisation-report/azure-file-service.js
Comment (JoshBinns2000) [reply portal-api/src/v1/services/utilisation-report/azure-file-service.js:1]: This looks fine to me

File: portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js
Comment (JoshBinns2000) [reply portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js:31]: Personally, I think the failing silently is something we want to add to the function being called rather than in the controller (a try-catch inside the try-catch feels a bit strange to read). My preferred method would be to have two methods exported from the emails service - one which can throw and has the name `sendUtilisation...OrThrow` and one which doesn't, but I'm not sure if that's the best solution here

**Review by JoshBinns2000 — CHANGES_REQUESTED** [review]
I think there are some places where prettier needs to be run, but otherwise looks pretty good

File: portal-api/src/v1/errors/invalid-report-status.error.ts
Comment (RLavender98) [reply portal-api/src/v1/errors/invalid-report-status.error.ts:1]: In the interest of this PR not taking on too many refactors I'm tempted to push against this for now

File: portal-api/src/v1/services/utilisation-report/email-service.ts
Comment (RLavender98) [reply portal-api/src/v1/services/utilisation-report/email-service.ts:24]: That was pre-existing, we get the formatted report period in the request from portal so that has been used I assume for simplicity over reformatting the report period object?

File: portal-api/src/v1/validation/utilisation-report/utilisation-report-upload-validator.ts
Comment (RLavender98) [reply portal-api/src/v1/validation/utilisation-report/utilisation-report-upload-validator.ts:5]: Yeah i struggled with this one a bit, I think I should perhaps move most of it back to the controller

File: portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js
Comment (RLavender98) [reply portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js:31]: Originally I had it in the downstream method but the issue I felt was then it doesn't match the behaviour of the other send email method, and also both send multiple emails and could throw sending one and not the others as well so 'OrThrow' seems a little misleading in and of itself if I add it to the other method...

File: portal-api/src/v1/api.js
Comment (JoshBinns2000) [reply portal-api/src/v1/api.js:359]: It looks like there may now be some indenting which is slightly off after removing the try-catch :)

File: portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js
Comment (JoshBinns2000) [inline portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js:16]: I think this may need to be `{import('@ukef/dtfs2-common').ReportPeriod}`?

File: portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js
Comment (JoshBinns2000) [inline portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js:49]: I think this is better than the function we had before - nice change!

File: portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js
Comment (JoshBinns2000) [inline portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js:59]: Do you think it is worthwhile for us to add a comment above this try-catch to explain why we don't want everything to fail if this throws?

File: portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js
Comment (JoshBinns2000) [inline portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js:71]: Could we update `InvalidReportStatusError` to have its own error code rather than hard-coding the `500` here?

File: portal-api/src/v1/services/utilisation-report/email-service.test.ts
Comment (JoshBinns2000) [inline portal-api/src/v1/services/utilisation-report/email-service.test.ts:52]: Do you think there is any benefit here to asserting that the error thrown is exactly what we expect rather than just checking that `Error` is thrown? I only ever include the `Error` constructor if I actually want to assert against the body of the error, like
```typescript
await expect(...).rejects.toThrow(new Error('my error message');
```
Otherwise, I tend to just use
```typescript
await expect(...).rejects.toThrow();
```
which I think does the same as the assertion we have here. Is there a benefit I'm missing with providing the plain `Error` object as an argument to `toThrow`?

File: portal-api/src/v1/validation/utilisation-report/utilisation-report-validator.test.ts
Comment (JoshBinns2000) [inline portal-api/src/v1/validation/utilisation-report/utilisation-report-validator.test.ts:21]: Should this be `'does not throw if the report has REPORT_NOT_RECEIVED status'`?

File: portal-api/src/v1/errors/invalid-report-status.error.ts
Comment (JoshBinns2000) [reply portal-api/src/v1/errors/invalid-report-status.error.ts:1]: That's fair. I left a comment about this elsewhere, but I think we should still include the specific http error code in this class as if it were using the `ApiError` class so that the controller doesn't need to hard code the `500` itself

File: portal-api/src/v1/services/utilisation-report/email-service.ts
Comment (JoshBinns2000) [reply portal-api/src/v1/services/utilisation-report/email-service.ts]: That makes sense to me

File: portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js
Comment (RLavender98) [reply portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js:59]: I guess I didn't add one because to me it feels fairly clear what the intention is and I am quite used to this being the pattern around downstream secondary actions. It is covered by a test case that you would have to intentionally remove to change this. That said, if this looks confusing to you as a reader then I will add a comment as that would be the only reason for a comment and as the author it is perhaps not down to me to solely judge that.

Comment (JoshBinns2000) [reply portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js:59]: It doesn't look confusing to me - I just wanted to know whether you thought it would be worth including a comment or not. More than happy to not include one as I do think it is quite clear as is :)

File: portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js
Comment (JoshBinns2000) [inline portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js:18]: Could this be `getReportByBankIdAndReportPeriod`?

File: portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js
Comment (JoshBinns2000) [inline portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js:32]: Missing semi-colon I think (now that the prettier PR has been merged, it should be possible to run `npm run prettier:fix` to fix any and all of these formatting things)

File: portal-api/src/v1/services/utilisation-report/email-service.test.ts
Comment (JoshBinns2000) [inline portal-api/src/v1/services/utilisation-report/email-service.test.ts:1]: This should import from `'../../api-response-types'` I think

**Review by JoshBinns2000 — COMMENTED** [review]
Only a couple of small comments - happy to approve once they're resolved and main is merged in

**Review by JoshBinns2000 — DISMISSED** [review]
LGTM

**Review by JoshBinns2000 — DISMISSED** [review]

**Review by JoshBinns2000 — APPROVED** [review]

**Pull request merged**
